bambu-studio: init at 01.06.02.04

zhaofengli opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

[BambuStudio](https://github.com/bambulab/BambuStudio) is a slicer for 3D printers made by Bambu Lab, based on PrusaSlicer.

~~This isn't ready to be merged yet. There are still missing functionalities in the Linux version (e.g., camera livestream) and the slicer sometimes crashes upon startup. Previously building the package with libraries shipped by Nixpkgs required [some invasive patches](https://github.com/bambulab/BambuStudio/pull/696) which have since been merged. Upstream has promised better Linux support so let's wait a bit.~~ (camera stream is now working and slicer no longer crashes intermittently)

~~The last commit is bandaid from #206221. I will remove it once the mpfr fixes land in master.~~ (done)

Ref: https://github.com/bambulab/BambuStudio/issues/12

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Moredread commented on pkgs/applications/misc/prusa-slicer/bambu-studio.nix:
```diff
@@ -0,0 +1,88 @@
+    };
+  } // args;
+
+  prusa-slicer' = prusa-slicer.override {
```
@cab404 who maintains super-slicer and me discussed somewhere (I can't find it ATM) that it makes sense to not use overrides for prusa-slicer anymore as there are quite a lot of differences now, i.e. super-slicer might be moved to an independent package, even if it involves some duplication. From the length of this file it looks like that might be a good idea for Bambo-Studio. What do you think?

zhaofengli replied:
Yeah, I agree and the build stage overrides indeed look very fragile. I'll rework this in a bit.

zhaofengli replied:
I've refactored the package to no longer be an override.

zhaofengli commented:
Thanks to the amazing work by @jwise in https://github.com/bambulab/BambuStudio/pull/1053 and https://github.com/bambulab/BambuStudio/pull/1045, we can use the EGL-enabled version of wxWidgets and the camera stream is now working. I've also added a fix for the intermittent crash and this seems ready to be merged.

jwise commented:
@zhaofengli What's the intermittent crash fix?  I know there's at least one I've been tracking.  Should we upstream it?

zhaofengli commented:
It's not actually a fix but kind of [an ugly workaround](https://github.com/NixOS/nixpkgs/pull/206495/files#diff-ae29f5f068573ef3e8480fd1f24c215b25005b32d27ae4f6951a807fa81ea456R151-R153).

For some reason, glew sometimes isn't initialized correctly and `__glewGenBuffers` is NULL. We link dynamically in nixpkgs - The upstream setup links statically which doesn't seem to cause issues.

It dies [here](https://github.com/bambulab/BambuStudio/blob/7e9ccefcaa2a275e7e8d0e0e25704e04275e12b1/src/slic3r/GUI/GLModel.cpp#L315) (see backtrace below):

```
   0x0000000000f1e6b9 <+25>:    push   %rbx
   0x0000000000f1e6ba <+26>:    mov    %rsi,%rbx
   0x0000000000f1e6bd <+29>:    lea    0x4(%rsi),%rsi
   0x0000000000f1e6c1 <+33>:    call   *(%r14)
=> 0x0000000000f1e6c4 <+36>:    mov    0xf7f4fd(%rip),%rbp        # 0x1e9dbc8
   0x0000000000f1e6cb <+43>:    mov    0x4(%rbx),%esi
   0x0000000000f1e6ce <+46>:    mov    $0x8892,%edi
(gdb) x $r14
0x7ffff0f12bf8 <__glewGenBuffers>:      0x00000000
```

<details>
<summary>Backtrace of segfault</summary>

At this point `wxGetApp().init_opengl()` was just called (see frame 8).

```
#0  0x0000000000000000 in ?? ()
#1  0x0000000000f1e6c4 in Slic3r::GUI::GLModel::send_to_gpu (this=<optimized out>, data=..., 
    vertices=std::vector of length 900, capacity 900 = {...}, indices=std::vector of length 150, capacity 150 = {...})
    at /build/source/src/slic3r/GUI/GLModel.cpp:315
#2  0x0000000000f1eb94 in Slic3r::GUI::GLModel::init_from (this=0x4603560, its=..., bbox=...)
    at /build/source/src/slic3r/GUI/GLModel.cpp:106
#3  0x0000000000f1f441 in Slic3r::GUI::GLModel::init_from (this=this@entry=0x4603560, its=...)
    at /build/source/src/slic3r/GUI/GLModel.cpp:112
#4  0x0000000000eb0faa in Slic3r::GUI::GLGizmoBase::GLGizmoBase (this=this@entry=0x4603490, parent=..., icon_filename="toolbar_move.svg", 
    sprite_id=sprite_id@entry=0) at /build/source/src/slic3r/GUI/Gizmos/GLGizmoBase.cpp:164
#5  0x0000000000eb24bf in Slic3r::GUI::GLGizmoMove3D::GLGizmoMove3D (this=this@entry=0x4603490, parent=..., 
    icon_filename="toolbar_move.svg", sprite_id=sprite_id@entry=0, obj_manipulation=obj_manipulation@entry=0x37ddfb0)
    at /build/source/src/slic3r/GUI/Gizmos/GLGizmoMove.cpp:32
#6  0x0000000000ea7365 in Slic3r::GUI::GLGizmosManager::init (this=this@entry=0x37ddda8)
    at /build/source/src/slic3r/GUI/Gizmos/GLGizmosManager.cpp:189
#7  0x0000000000e7503c in Slic3r::GUI::GLCanvas3D::init (this=0x37ddc00) at /build/source/src/slic3r/GUI/GLCanvas3D.cpp:1220
#8  0x0000000000fa2202 in Slic3r::GUI::GUI_App::post_init (this=0x24505e0) at /build/source/src/slic3r/GUI/GUI_App.cpp:1025
#9  0x00007ffff0d30e96 in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#10 0x00007ffff0d311d6 in wxEvtHandler::SearchDynamicEventTable(wxEvent&) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#11 0x00007ffff0d31340 in wxEvtHandler::TryHereOnly(wxEvent&) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#12 0x00007ffff0d313eb in wxEvtHandler::ProcessEventLocally(wxEvent&) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#13 0x00007ffff0d31491 in wxEvtHandler::ProcessEvent(wxEvent&) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#14 0x00007ffff0c150f8 in wxAppConsoleBase::ProcessIdle() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#15 0x00007ffff07d7724 in wxAppBase::ProcessIdle() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_gtk3u_core-3.1.so.5
#16 0x00007ffff0717aa9 in wxApp::DoIdle() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_gtk3u_core-3.1.so.5
#17 0x00007ffff0717b83 in wxapp_idle_callback ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_gtk3u_core-3.1.so.5
#18 0x00007fffee06a8e4 in g_main_context_dispatch () from /nix/store/61qy7d70070djqcg135l22v18a5py758-glib-2.74.3/lib/libglib-2.0.so.0
#19 0x00007fffee06ac88 in g_main_context_iterate.constprop ()
   from /nix/store/61qy7d70070djqcg135l22v18a5py758-glib-2.74.3/lib/libglib-2.0.so.0
#20 0x00007fffee06af73 in g_main_loop_run () from /nix/store/61qy7d70070djqcg135l22v18a5py758-glib-2.74.3/lib/libglib-2.0.so.0
#21 0x00007fffee9c23c5 in gtk_main () from /nix/store/rv6nydaizz3n8s3l3jv6pss0pzxn3yzm-gtk+3-3.24.35/lib/libgtk-3.so.0
#22 0x00007ffff0727c65 in wxGUIEventLoop::DoRun() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_gtk3u_core-3.1.so.5
#23 0x00007ffff0c3f37d in wxEventLoopBase::Run() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#24 0x00007ffff0c1491b in wxAppConsoleBase::MainLoop() ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#25 0x00007ffff0c81eb6 in wxEntry(int&, wchar_t**) ()
   from /nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5/lib/libwx_baseu-3.1.so.5
#26 0x0000000000f6e46b in Slic3r::GUI::GUI_Run (params=...) at /build/source/src/slic3r/GUI/GUI_Init.cpp:64
#27 0x00000000007a8f1b in Slic3r::CLI::run (this=0x7fffffffb9c0, argc=<optimized out>, argv=<optimized out>)
    at /build/source/src/BambuStudio.cpp:452
#28 0x0000000000763255 in main (argc=<optimized out>, argv=<optimized out>) at /build/source/src/BambuStudio.cpp:2625
```
</summary>

jwise commented:
Oh, this is the crash where things go wrong if the window doesn't have focus?

You definitely should not be linking dynamically against an external GLEW, unless you are specifying that it is an EGL GLEW (or that the EGLness of the GLEW, anyway, is the same as the EGLness of the wxWidgets).  GLEW can either be built as EGL or GLX, but not both at the same time.  It is much better to use the vendored GLEW, for this reason.

Note also that if you are linking against an external wxWidgets, there are at least some BBS-specific patches to wxWidgets in their tree, too, which is why they build that statically...

zhaofengli commented:
This package is indeed [specified](https://github.com/NixOS/nixpkgs/pull/206495/files#diff-ab5748dc9567516fefba8344056b51ec1866adeace380f46e58a7af3d619ea22R32483-R32492) to use the EGL version of GLEW and a special version of wxWidgets with WebRequests and private fonts enabled.[^1] When you specify a dependency, the resulting executables have their runpaths pinned to that specific version of dependency (e.g., `/nix/store/83gnq2f6h1jw2n7fndjjfgf50gf4894i-wxwidgets-3.1.5`). Everything is pinned and global libraries simply do not exist (no `/lib` or `/usr/lib`).

You can `curl.overrideAttrs ({ patches ? [], ... } { patches = patches ++ [ ./my.patch ]; })` and it evaluates to a patched version of curl that gets installed to a path that doesn't affect any other curl in the system. It's all functional and I invite you to give it a try :smile: 

Because each build happens in a clean sandbox without network access, it's difficult to use the vendored deps and we generally avoid doing so unless necessary. Regarding the wxWidgets patches, I don't immediately see a need from a quick glance but if we do, we _can_ easily add them (see example above).

[^1]: Yes, I do need to remove the commented out `withEGL = false` override...

norpol commented:
I'm running on latest nixos-unstable and I'm building/launching bambu-studio via

```bash
nix shell github:zhaofengli/nixpkgs/bambu-studio#bambu-studio
bambu-studio
```

Which starts-up (flashes open on my screen) and then crashes in the same moment. 

As you might see on the video recording, you notice that the UI almost presents completely up until around these errors:
[screencast.webm](https://user-images.githubusercontent.com/2230917/220697235-01e7888e-93f2-4646-a2a9-ba7887237643.webm)


```
(bambu-studio:8781): Gtk-CRITICAL **: 16:38:44.084: gtk_widget_set_size_request: assertion 'width >= -1' failed
libEGL warning: MESA-LOADER: failed to open radeonsi: /nix/store/wprxx5zkkk13hpj6k1v6qadjylh3vq9m-gcc-11.3.0-lib/lib/libstdc++.so.6: version
`GLIBCXX_3.4.30' not found (required by /nix/store/yvcnhajqbwcq35vpnsm1waw557dxcn5v-llvm-15.0.7-lib/lib/libLLVM-15.so) (search paths /run/ope
ngl-driver/lib/dri, suffix _dri)

libEGL warning: MESA-LOADER: failed to open swrast: /nix/store/wprxx5zkkk13hpj6k1v6qadjylh3vq9m-gcc-11.3.0-lib/lib/libstdc++.so.6: version `G
LIBCXX_3.4.30' not found (required by /nix/store/yvcnhajqbwcq35vpnsm1waw557dxcn5v-llvm-15.0.7-lib/lib/libLLVM-15.so) (search paths /run/openg
l-driver/lib/dri, suffix _dri)

EGLDisplay Initialization failed: EGL_NOT_INITIALIZED
Cannot create EGL context: invalid display (last error: EGL_SUCCESS)

(bambu-studio:8781): Gtk-CRITICAL **: 16:38:44.229: gtk_gesture_zoom_new: assertion 'GTK_IS_WIDGET (widget)' failed
```

norpol commented:
I've rebased your branch to latest nixos-unstable and added some [gcc-unwrapped as a dependency](https://github.com/NixOS/nixpkgs/commit/b91bd7dfe19b4c02e42d70299ebc885b60f1a9fb) which is [similarish to here](https://github.com/NixOS/nixpkgs/pull/218665/files), not sure if that was the cause but it builds now. 

Also since they've just released the 1.5 Beta version, I gave building this a shot and it seems to be working.

Only issue right now is the error "Media playback error: Could not open audio device for playback." when I'm trying to start the webcam.

![2023-03-02-151449](https://user-images.githubusercontent.com/2230917/222470242-45df744a-92b5-4ffd-ac76-d3529648155d.jpeg)

zhaofengli commented:
> I've rebased your branch to latest nixos-unstable and added some [gcc-unwrapped as a dependency](https://github.com/NixOS/nixpkgs/commit/b91bd7dfe19b4c02e42d70299ebc885b60f1a9fb) which is [similarish to here](https://github.com/NixOS/nixpkgs/pull/218665/files), not sure if that was the cause but it builds now.
> 
> Also since they've just released the 1.5 Beta version, I gave building this a shot and it seems to be working. [86286f4](https://github.com/NixOS/nixpkgs/commit/86286f41b01d6901e335fece713c6dd392f62c71)

Thanks! I've added the fix to the PR and bumped the version.

> Only issue right now is the error "Media playback error: Could not open audio device for playback." when I'm trying to start the webcam.

Hmm, everything seems to be working fine for me with Pipewire.

norpol commented:
I've also built https://github.com/SoftFever/BambuStudio-SoftFever, which supports other printers + several calibration gcode generators.

Might also be worth  packaging the @SoftFever version as it looks quite useful.

There is a known bug showing up, that's also affecting their app-image version and I've created a separate report for that https://github.com/SoftFever/BambuStudio-SoftFever/issues/442


```patch
diff --git a/pkgs/applications/misc/bambu-studio/default.nix b/pkgs/applications/misc/bambu-studio/default.nix
index 5ffceb22fd..acb2a0fb8b 100644
--- a/pkgs/applications/misc/bambu-studio/default.nix
+++ b/pkgs/applications/misc/bambu-studio/default.nix
@@ -50,14 +50,15 @@ let
   });
 in
 stdenv.mkDerivation rec {
-  pname = "bambu-studio";
-  version = "01.05.00.59";
+  pname = "bambu-studio-softfever";
+  version = "1.4.5";

   src = fetchFromGitHub {
-    owner = "bambulab";
-    repo = "BambuStudio";
+    owner = "softfever";
+    repo = "BambuStudio-SoftFever";
     rev = "v${version}";
-    hash = "sha256-NCd2hT6xVaTySYkacAV1wT1fOe8yq4qPxmEzUdVHEDw=";
+    hash = "sha256-jwtYzcDQb3LpABf8btgiKRuLPUMhrDTRE/lqtaTgGU4=";
   };

   nativeBuildInputs = [
@@ -158,7 +159,7 @@ stdenv.mkDerivation rec {

   meta = with lib; {
     description = "PC Software for BambuLab's 3D printers";
-    homepage = "https://github.com/bambulab/BambuStudio";
+    homepage = "https://github.com/SoftFever/BambuStudio-SoftFever";
     license = licenses.agpl3;
     maintainers = with maintainers; [ zhaofengli ];
     mainProgram = "bambu-studio";
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 255e4b0a69..c1528d4f7d 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -33244,7 +33244,7 @@ with pkgs;

   super-slicer-latest = super-slicer.latest;

-  bambu-studio = callPackage ../applications/misc/bambu-studio {
+  bambu-studio-softfever = callPackage ../applications/misc/bambu-studio {
     inherit (gst_all_1) gstreamer gst-plugins-base gst-plugins-bad;

     # Needs to be the same version as in the closed-source networking plugin.
```

deftdawg commented:
Random chime in...   (I'm a nixOS noob, so this may not be helpful)

I installed this on `22.11` via:
```sh
export REV=6b1493ed77d54947f043d30f1960ed343d90ca50
nix --extra-experimental-features flakes profile install --impure github:zhaofengli/nixpkgs/${REV}#bambu-studio

# Run it with LC_ALL or it will crash
LC_ALL=C bambu-studio
```

It starts to init, and then coredumps...   maybe it's because I'm running X11 instead of Wayland or something, could also be upstream bug... anywhere here's a backtrace
```cpp
Thread 40 "bambu-stud:sh0" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff57fff640 (LWP 253620)]
0x00007fffde493b40 in llvm::MachinePointerInfo::getAddrSpace() const () from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
(gdb) bt
#0  0x00007fffde493b40 in llvm::MachinePointerInfo::getAddrSpace() const () from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
#1  0x00007fffde9c0a0a in AddNodeIDCustom(llvm::FoldingSetNodeID&, llvm::SDNode const*) () from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
#2  0x00007fffde9c0f54 in llvm::FoldingSet<llvm::SDNode>::NodeEquals(llvm::FoldingSetBase const*, llvm::FoldingSetBase::Node*, llvm::FoldingSetNodeID const&, unsigned int, llvm::FoldingSetNodeID&) ()
   from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
#3  0x00007fff5d9c09ab in llvm::FoldingSetBase::FindNodeOrInsertPos(llvm::FoldingSetNodeID const&, void*&, llvm::FoldingSetBase::FoldingSetInfo const&) ()
   from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#4  0x00007fff5d9c0a90 in llvm::FoldingSetBase::GetOrInsertNode(llvm::FoldingSetBase::Node*, llvm::FoldingSetBase::FoldingSetInfo const&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#5  0x00007fff5e4316e3 in llvm::SelectionDAG::AddModifiedNodeToCSEMaps(llvm::SDNode*) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#6  0x00007fff5e431eaa in llvm::SelectionDAG::ReplaceAllUsesWith(llvm::SDValue, llvm::SDValue) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#7  0x00007fff5e2dbcec in (anonymous namespace)::SelectionDAGLegalize::ReplaceNode(llvm::SDValue, llvm::SDValue) [clone .constprop.0] () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#8  0x00007fff5e2f35b5 in llvm::SelectionDAG::Legalize() () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#9  0x00007fff5e464af3 in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#10 0x00007fff5e4677fc in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#11 0x00007fff5e46a012 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) [clone .part.0] () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#12 0x00007fff5def2d97 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#13 0x00007fff5dc1fc30 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#14 0x00007fff5f267eca in (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#15 0x00007fff5dc21634 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /nix/store/gaamqa1p6ql40y80d18rpd2q6xi6w7hn-llvm-14.0.6-lib/lib/libLLVM-14.so
#16 0x00007fff712e2206 in ac_compile_module_to_elf () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#17 0x00007fff712336e1 in si_compile_llvm () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#18 0x00007fff71235e9c in si_llvm_compile_shader () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#19 0x00007fff7123024e in si_compile_shader () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#20 0x00007fff71256e1d in si_init_shader_selector_async(void*, void*, int) () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#21 0x00007fff70ab7099 in util_queue_thread_func () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#22 0x00007fff70b020c7 in impl_thrd_routine () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#23 0x00007fffe9688e86 in start_thread () from /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6
#24 0x00007fffe970fd30 in clone3 () from /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6
```

norpol commented:
@deftdawg Try the nix shell I've provided. Also (I'm not sure) but might be an issue with the mismatch between unstable vs 22.11. Also might be related to an outdated nix profile? At least back when `nix-env -iA` was still a thing you'd run into similar issues frequently.

deftdawg commented:
> @deftdawg Try the nix shell I've provided (Sorry might rebuild everything). Also (I'm not sure) but might be an issue with the mismatch between unstable vs 22.11. Also might be related to an outdated nix profile? At least back when `nix-env -iA` was still a thing you'd run into similar issues frequently.
> 
> ```
> nix shell github:zhaofengli/nixpkgs/bambu-studio#bambu-studio
> bambu-studio
> ```

Thanks @norpol.  Result is the same, it's probably something with my setup...

The nix store artifact is the same when I build it through `nix install profile` or `nix shell`, maybe because it's cached... 

In both cases it uses this:
`/nix/store/ppiyw7yyfgh67yfb83z1ay9jcy1gymyx-bambu-studio-01.05.00.59/bin/bambu-studio`

Is that the same hash you have for it?

norpol commented:
@deftdawg `/nix/store/ppiyw7yyfgh67yfb83z1ay9jcy1gymyx-bambu-studio-01.05.00.59/` same path indeed.

deftdawg commented:
@norpol here's the `LC_ALL=C strace -f -s1024 .bambu-studio-wrapped 2&> ~/bambu-studio-playbin-strace.txt`

[bambu-studio-playbin-strace.txt.gz](https://github.com/NixOS/nixpkgs/files/11035089/bambu-studio-playbin-strace.txt.gz)

When I run the wrapper, it paints the following...  perhaps it can't find gstreamer or the network plugin.
![Screenshot_20230321_202022](https://user-images.githubusercontent.com/4991612/226770740-4735e321-9d71-4518-acda-659527dece4a.png)

deftdawg commented:
I built Zhaofengli's older `bambu-studio-old` tag 

```
export REV=b9f18ffc3489fc926843a4b64d9fae79ddee8aa8
nix --extra-experimental-features flakes profile install --impure github:zhaofengli/nixpkgs/${REV}#bambu-studio
bambu-studio 
```

It still gives the same error message about `invalid playbin`, however it does not segfault...  So I'm using that for now.

zhaofengli commented:
Just bumped the version to [01.05.00.61](https://github.com/bambulab/BambuStudio/releases/tag/v01.05.00.61) ([diff](https://github.com/bambulab/BambuStudio/compare/v01.05.00.59...v01.05.00.61)) and it works for me. We should look deeper into why it's segfaulting for some people.

deftdawg commented:
Newest build dies with same backtrace...  Since it says radeon (lines 13-19), makes me wonder what gpus @zhaofengli and @norpol are running?   AMD (me) / Nvidia / Intel?

```sh
export REV=e2f3ffe9d593412fd48bbf4ed382e4b5f219a523
time nix --extra-experimental-features flakes profile install --impure github:zhaofengli/nixpkgs/${REV}#bambu-studio
echo "bt" | coredumpctl debug # dump backtrace of last core
```

```cpp
(gdb) bt
#0  0x00007f2a78693b40 in llvm::MachinePointerInfo::getAddrSpace() const () from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
#1  0x00007f2a78bc0a0a in AddNodeIDCustom(llvm::FoldingSetNodeID&, llvm::SDNode const*) () from /nix/store/1a31x70as21dm8dbl25h1qqvy5if4qsi-llvm-15.0.7-lib/lib/libLLVM-15.so
#2  0x00007f28e99c0a7d in llvm::FoldingSetBase::GetOrInsertNode(llvm::FoldingSetBase::Node*, llvm::FoldingSetBase::FoldingSetInfo const&) ()
   from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#3  0x00007f28ea4316e3 in llvm::SelectionDAG::AddModifiedNodeToCSEMaps(llvm::SDNode*) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#4  0x00007f28ea4314e0 in llvm::SelectionDAG::ReplaceAllUsesWith(llvm::SDNode*, llvm::SDNode*) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#5  0x00007f28ea2b0c78 in llvm::SelectionDAG::Combine(llvm::CombineLevel, llvm::AAResults*, llvm::CodeGenOpt::Level) ()
   from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#6  0x00007f28ea464b58 in llvm::SelectionDAGISel::CodeGenAndEmitDAG() () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#7  0x00007f28ea4677fc in llvm::SelectionDAGISel::SelectAllBasicBlocks(llvm::Function const&) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#8  0x00007f28ea46a012 in llvm::SelectionDAGISel::runOnMachineFunction(llvm::MachineFunction&) [clone .part.0] ()
   from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#9  0x00007f28e9ef2d97 in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#10 0x00007f28e9c1fc30 in llvm::FPPassManager::runOnFunction(llvm::Function&) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#11 0x00007f28eb267eca in (anonymous namespace)::CGPassManager::runOnModule(llvm::Module&) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#12 0x00007f28e9c21634 in llvm::legacy::PassManagerImpl::run(llvm::Module&) () from /nix/store/np4sh6in90c2yjm5znfr1qcaznlmyj7l-llvm-14.0.6-lib/lib/libLLVM-14.so
#13 0x00007f2a022e2206 in ac_compile_module_to_elf () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#14 0x00007f2a022336e1 in si_compile_llvm () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#15 0x00007f2a02235e9c in si_llvm_compile_shader () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#16 0x00007f2a0223024e in si_compile_shader () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#17 0x00007f2a02256e1d in si_init_shader_selector_async(void*, void*, int) () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#18 0x00007f2a01ab7099 in util_queue_thread_func () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#19 0x00007f2a01b020c7 in impl_thrd_routine () from /run/opengl-driver/lib/dri/radeonsi_dri.so
#20 0x00007f2a83888e86 in start_thread () from /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6
#21 0x00007f2a8390fd30 in clone3 () from /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6
(gdb) 
```

zhaofengli commented:
I'm using i915 (Arc A770).

> /nix/store/bfbp3ypd9nm3fapz634gvvs738blrl0y-gcc-wrapper-12.2.0/bin/ar: unable to copy file 'liblibslic3r_gui.a'; reason: Success

Hmm, weird. Maybe something got OOM-killed and it was confused.

norpol commented:
> Hmm, weird. Maybe something got OOM-killed and it was confused.

Tried building it several times until I received a different error `ar: reason: No space left on device` which gave me the clue that I forgot that I changed `/tmp` a `tmpfs` which apparently is too small with 4GB (and memory pressure?) for the build. My build with a bigger disk tmpfs is successful and startup works too. Thank you @zhaofengli :)

deftdawg commented:
Bumped to newest release 01.06.02.04 (1.6.2 beta) and rebuilt, it still crashes out:
```patch
diff --git a/pkgs/applications/misc/bambu-studio/default.nix b/pkgs/applications/misc/bambu-studio/default.nix
index bab2a82c38e..4925f23a9a0 100644
--- a/pkgs/applications/misc/bambu-studio/default.nix
+++ b/pkgs/applications/misc/bambu-studio/default.nix
@@ -51,13 +51,13 @@ let
 in
 stdenv.mkDerivation rec {
   pname = "bambu-studio";
-  version = "01.05.00.61";
+  version = "01.06.02.04";
 
   src = fetchFromGitHub {
     owner = "bambulab";
     repo = "BambuStudio";
     rev = "v${version}";
-    hash = "sha256-ixZaU0Q8yI/z624KX6mbpZ7y0IsA5NMDadLmPHyGL64=";
+    hash = "sha256-k2/0ukIVjGhUUKhaJCldTRsQcHwX/mL2ROVaiTkv/eg=";
   };
 
   nativeBuildInputs = [
```

Now also seeing  `Can't find variable: __wxOut0` before it core dumps - that's an upstream issue: https://github.com/bambulab/BambuStudio/issues/1701

balacij commented:
I don't seem to be able to log in to Bambu Studio with this.

I'm getting this error followed by an HTTP request dump:
```console

(bambu-studio:81108): Gtk-CRITICAL **: 22:21:40.235: gtk_cell_layout_get_cells: assertion 'GTK_IS_CELL_LAYOUT (cell_layout)' failed

** (bambu-studio:81108): CRITICAL **: 22:21:40.245: Cannot register URI scheme wxfs more than once

** (bambu-studio:81108): CRITICAL **: 22:21:40.245: Cannot register URI scheme memory more than once
/nix/store/kw3qkcg847yzrkaw0llnzlmkcih1mpls-gvfs-1.50.4/lib/gvfs/libgvfscommon.so: undefined symbol: g_task_set_static_name
Failed to load module: /nix/store/kw3qkcg847yzrkaw0llnzlmkcih1mpls-gvfs-1.50.4/lib/gio/modules/libgvfsdbus.so
libEGL warning: MESA-LOADER: failed to open radeonsi: /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6: version `GLIBC_2.36' not found (required by /nix/store/98sq0hkrbda37k9iayqdqchkibp8530y-llvm-15.0.7-lib/lib/libLLVM-15.so) (search paths /run/opengl-driver/lib/dri, suffix _dri)

libEGL warning: MESA-LOADER: failed to open swrast: /nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224/lib/libc.so.6: version `GLIBC_2.36' not found (required by /nix/store/98sq0hkrbda37k9iayqdqchkibp8530y-llvm-15.0.7-lib/lib/libLLVM-15.so) (search paths /run/opengl-driver/lib/dri, suffix _dri)

EGLDisplay Initialization failed: EGL_NOT_INITIALIZED
Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
*   Trying 54.153.50.31:443...
* Connected to e.bambulab.com (54.153.50.31) port 443 (#0)
* ALPN: offers h2,http/1.1
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=*.bambulab.com
*  start date: Jun 30 00:00:00 2022 GMT
*  expire date: Jul 15 23:59:59 2023 GMT
*  issuer: C=US; O=DigiCert, Inc.; CN=RapidSSL Global TLS RSA4096 SHA256 2022 CA1
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* using HTTP/2
...
```

zhaofengli commented:
With #232521 merged, now the package can no longer be evaluated by default on top of master because the closed-source network plugin depends on OpenSSL 1.1 :disappointed: I'm thinking about opening an issue, but I guess it'll take time on their end.

@balacij There is likely a mismatch between the nixpkgs version your system is built on and the nixpkgs version in this PR. In the meantime, I'll probably create a flake with an overlay in a separate repo so this can be easily used.

deftdawg commented:
> I'm thinking about opening an issue, but I guess it'll take time on their end.

Done :smile:

deftdawg commented:
Some good news, with the exception of the camera, this now builds and runs for me! 
```
NIXPKGS_ALLOW_INSECURE=1 nix-build -A bambu-studio && result/bin/bambu-studio
```

Not sure if it's updating to 23.05 or just the upstream repo changes you just pushed...  But it's good good :+1: :+1:

zhaofengli commented:
> Some good news, with the exception of the camera, this now builds and runs for me!

The camera works for me at least. Any logs with `--debug 5`?

There is [a tiny patch](https://github.com/bambulab/BambuStudio/pull/1906) to fix the `wxTranslations::Get()` segfault, but otherwise it's a simple bump.

zhaofengli committed:
wxGTK31: Support enabling private fonts

zhaofengli committed:
bambu-studio: init at 01.06.02.04

Co-authored-by: Phileas Lebada <gitpl@woerm.at>

zhaofengli committed:
wxGTK31: Support enabling WebRequest component

JeremiahSecrist commented:
Wonderful work so far, guys!
Can anyone else see if they run into the same issue? The application segfaults when doing the steps below.
Let me know if you want any more info.
**Steps:**
```bash
export NIXPKGS_ALLOW_INSECURE=1 export REV=fb3cca0aa1f240de37da7a3567b38b2961416394
nix profile install --impure github:zhaofengli/nixpkgs/${REV}#bambu-studio
```
- Sign into bambu studio
- try to edit filament
- witness segfault

zhaofengli committed:
bambu-studio: 01.06.02.04 -> 01.07.00.65

noisersup commented:
@JeremiahSecrist couldn't reproduce your bug on my machine on the latest commit. I will try to test it with a printer in following days.

@zhaofengli Does anything need to be addressed to finish this PR? Also, thanks for the great job on it!

aacebedo commented:
Hey great work! If it is ready can we push it in the discourse to get it merged?

JeremiahSecrist commented:
@noisersup maybe it is Wayland specific?

ovlach commented:
> @noisersup maybe it is Wayland specific?

Don't known if this case, but OrcaSlicer (fork of Bambu studo) give segv after i try edit filament (before edit dialog shows)... I trace-it and this patch "fixed it" -- https://github.com/ovlach/nix-orca-slicer/blob/main/0001-segv-patches.patch. I don't have time to deeply analyze, so i only comment lines...

cab404 commented:
Won't build on master via nixpkgs-review due to openssl 1.1 being EOL and emitted warnings~

noisersup commented:
Just testing it with bambu lab X1 carbon. I was able to easily detect the printer, setup filament, slice and print the file. I encountered the segfault while trying to open camera though.

After updating package to `v01.07.01.62` there's no segfault, although it shows only the first frame of the stream.

JeremiahSecrist commented:
@ovlach That is the issue I'm getting! I wonder why no one else here is running into it.
`(bambu-studio:8077): Gtk-CRITICAL **: 22:22:33.992: gtk_window_resize: assertion 'height > 0' failed`

tomberek commented:
Is this perhaps related to submodules? https://github.com/bambulab/BambuStudio/blob/master/BuildLinux.sh#L144C10-L144C18

Mastermindaxe commented:
Is there any way I can help move this along without super extensive knowledge of nix? If I can test some stuff for you guys or sth pls let me know. I'm dying to have this in the upstream nixpkgs

MathiasSven commented:
> Is there any way I can help move this along without super extensive knowledge of nix? If I can test some stuff for you guys or sth pls let me know. I'm dying to have this in the upstream nixpkgs

The main holdover right now is the upstream's version of OpenSSL being EOL. It seems that it will be soon addressed though. https://github.com/bambulab/BambuStudio/issues/1843#issuecomment-1678483485

JeremyKennedy commented:
Hey all, I am brand new to NixOS and wanted to get the latest BambuStudio version running. I had to fork @zhaofengli's repo and make some changes for it to pull the latest version. For anyone else in the same boat, here's how to run it using my changes:
```
export NIXPKGS_ALLOW_INSECURE=1
export NIXPKGS_ALLOW_UNFREE=1
nix run "github:JeremyKennedy/nixpkgs?ref=40a6c60e6cc97b0d5db17a20f9953c9dbf2e0ea5#bambu-studio" --impure
```

And for anyone looking to update to a future version (in case myself or @zhaofengli don't make the change in time):
1. Update BambuStudio's [default.nix](https://github.com/JeremyKennedy/nixpkgs/blob/40a6c60e6cc97b0d5db17a20f9953c9dbf2e0ea5/pkgs/applications/misc/bambu-studio/default.nix#L54) with the new version found on [the releases page](https://github.com/bambulab/BambuStudio/releases), and the hash using the `hash` output from the commands below, where `COMMIT` is the commit found on the releases page.
```
nix-shell -p nix-prefetch-git
nix-prefetch-git http://github.com/bambulab/BambuStudio COMMIT
```

2. Commit these changes to your own repo, then update the `nix run` command above with the commit hash from your commit.

Apologies if these steps are not detailed enough, I just wanted to quickly get the info out there for anyone else.

aacebedo commented:
Would it be possible to also integrate this patch?
https://github.com/NixOS/nixpkgs/pull/206495#issuecomment-1674364131

I cannot edit the filament settings without a crash and it seems this solves the issue.

deftdawg commented:
It looks like all the dev activity for 1.07.06.92 has been occurring on the[ v1.7.2. branch](https://github.com/bambulab/BambuStudio/commits/v1.7.2), the [release notes for v01.07.06.92](https://github.com/bambulab/BambuStudio/releases/tag/v01.07.06.92) suggest OpenSSL has been updated to 3.1.

zhaofengli committed:
bambu-studio: 01.07.00.65 -> 01.07.06.92

The OpenSSL pin is no longer needed.

norpol commented:
Thanks @aacebedo reported it over at BambuLab https://github.com/bambulab/BambuStudio/issues/2605 and applied the patch for myself and verified this workaround is working at least to get the window open.
Thanks @zhaofengli for bumping, did the same last night and can very build working. Updated the Patch
Hope my patched git format-patch works, otherwise you'll have to copy the lines sorry. 

```path
From 9b54fe3e9eba2ad4fbb9248aa4e713d50634272d Mon Sep 17 00:00:00 2001
From: Phileas Lebada <invalid.invalid>
Date: Mon, 25 Sep 2023 00:00:26 +0100
Subject: [PATCH] bambu-studio: 01.07.00.65 -> 01.07.06.92 and apply workaround
 filament segvault

---
 .../misc/bambu-studio/0002-segv-patches.patch   | 17 +++++++++++++++++
 pkgs/applications/misc/bambu-studio/default.nix |  2 ++
 2 files changed, 19 insertions(+)
 create mode 100644 pkgs/applications/misc/bambu-studio/0002-segv-patches.patch

diff --git a/pkgs/applications/misc/bambu-studio/0002-segv-patches.patch b/pkgs/applications/misc/bambu-studio/0002-segv-patches.patch
new file mode 100644
index 0000000000..6aef942da3
--- /dev/null
+++ b/pkgs/applications/misc/bambu-studio/0002-segv-patches.patch
@@ -0,0 +1,17 @@
+diff -ruN ./src/slic3r/GUI/Field.cpp ../../source/src/slic3r/GUI/Field.cpp
+--- ./src/slic3r/GUI/Field.cpp	1970-01-01 01:00:01.000000000 +0100
++++ ../../source/src/slic3r/GUI/Field.cpp	2023-07-20 01:19:02.755757640 +0200
+--- oldfile.cpp	2023-09-24 23:56:14.054014696 +0100
++++ newfile.cpp	2023-09-24 23:58:20.086430254 +0100
+@@ -1541,8 +1541,9 @@
+ 
+     wxButton* btn = dynamic_cast<wxButton*>(field->GetPickerCtrl());
+     wxImage image(btn->GetBitmap().GetSize());
+-    image.InitAlpha();
+-    memset(image.GetAlpha(), 0, image.GetWidth() * image.GetHeight());
++    // TODO: why initAlpha() raise segv?
++    /*image.InitAlpha();
++    memset(image.GetAlpha(), 0, image.GetWidth() * image.GetHeight());*/
+     wxBitmap   bmp(std::move(image));
+     wxMemoryDC dc(bmp);
+     if (!dc.IsOk()) return;
diff --git a/pkgs/applications/misc/bambu-studio/default.nix b/pkgs/applications/misc/bambu-studio/default.nix
index 191aa4659a..94d10e8d8a 100644
--- a/pkgs/applications/misc/bambu-studio/default.nix
+++ b/pkgs/applications/misc/bambu-studio/default.nix
@@ -105,6 +105,8 @@ stdenv.mkDerivation rec {
   patches = [
     # Fix for webkitgtk linking
     ./0001-not-for-upstream-CMakeLists-Link-against-webkit2gtk-.patch
+    # Fix from orca-slicer filament profile editor segvault crash workaround
+    ./0002-segv-patches.patch
   ];
 
   doCheck = true;
-- 
2.42.0
```

aacebedo commented:
I updated my system to unstable 2 days ago and I can't start bambu-studio anymore (it segfaults immediately). Is anyone having the same problem?

aacebedo commented:
> [1.07.06.92](https://github.com/bambulab/BambuStudio/releases/tag/v01.07.06.92) s

I tried 01.07.06.92 same issue. Plus the fact that on Hyprland it just closes itself immediately (no segfault).

MathiasSven commented:
I updated my system a couple of days ago, I hand't tried to start bambu-studio since, just tried it now, and I am also getting a segmentation fault. Although I am still on `BambuStudio-01.07.00.65`.

aacebedo commented:
> I updated my system a couple of days ago, I hand't tried to start bambu-studio since, just tried it now, and I am also getting a segmentation fault. Although I am still on `BambuStudio-01.07.00.65`.

I tried 01.07.06.92 same issue. Plus the fact that on Hyprland it just closes itself immediately (no segfault).

noisersup commented:
The same segfault issue after the update, both on wayland, and x11.

aacebedo commented:
> The same segfault issue after the update, both on wayland, and x11.

Also tried the other PR for orca slicer. Same issue.

noisersup commented:
Okay, got some time and tried to obtain a bit more information.

```
Cannot get default EGL display: EGL_BAD_PARAMETER
Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
Cannot get default EGL display: EGL_BAD_PARAMETER
Cannot get default EGL display: EGL_BAD_PARAMETER
Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
```

Those logs print out in my latest (updated) NixOS generation, while the old, working one doesn't print those. Maybe it's related to Glib which bumped version in the meantime?

Commit:
bambu-studio: 01.07.06.92 -> 01.08.00.62

* Switch to newer tbb version

tbb was broken up into two versions for backwards compability

https://github.com/NixOS/nixpkgs/pull/217585/files

and put into prusa-slicer

https://github.com/NixOS/nixpkgs/pull/238921/files

Which fixes for 1.8 beta build
```
bambu-studio> CMake Error at src/libslic3r/CMakeLists.txt:490 (target_link_libraries):
bambu-studio>   Target "libslic3r" links to:
bambu-studio>     TBB::tbbmalloc
bambu-studio>   but the target was not found.  Possible reasons include:
bambu-studio>     * There is a typo in the target name.
bambu-studio>     * A find_package call is missing for an IMPORTED target.
bambu-studio>     * An ALIAS target is missing.
bambu-studio>
bambu-studio> -- Generating done (0.2s)
bambu-studio> CMake Warning:
bambu-studio>   Manually-specified variables were not used by the project:
bambu-studio>     CMAKE_EXPORT_NO_PACKAGE_REGISTRY
bambu-studio>     DEP_WX_GTK3
bambu-studio>
bambu-studio> CMake Generate step failed.  Build files cannot be regenerated correctly.
error: builder for '/nix/store/rzgx01p1ni2xbx5ddg0133pjjlzs8618-bambu-studio-01.08.00.57.drv' failed with exit code 1;
```

* bambu-studio: switch to custom openvdb tbb version

* Downgrade boost to 179 (181 is current default)

norpol commented:
Yes, it's likely the glibc version. Starting with `--debug 5` will show some glibc related error messages. I've rebased and switched to the latest beta release (1.8.0), but that now fails because of changes in `tbb` and `wxGTK`. Starting point might be `prusa-slicer` which incorporates some patches to support `wxGTK32` patches.

With `glibc` updated, `tbb` newer and `openvdb` related to that I get till ([diff](https://github.com/zhaofengli/nixpkgs/pull/2/files/a0dbc34dfe422fbefff8d6c368047d0c18ca9ab1..b928acef50b7acf3d2731edbe5d8a0016bd29986)).

Found also another issue with the same error https://github.com/prusa3d/PrusaSlicer/issues/9294

which probably talks about this [gentoo prusaslicer boost patch](https://github.com/gentoo/gentoo/blob/7502e4c9bbfca3d435ae45d900efa56ea0b9a6a5/media-gfx/prusaslicer/files/prusaslicer-2.5.0_alpha2-boost-fixes.patch)

```
include/boost/filesystem/detail/path_traits.hpp:495: error: call of overloaded '_check_convertible_to_path_source(const wxCStrData&)' is ambiguous
```

I'm not sure if just that patch will do the job, the prusa-slicer has no boost patch but instead patches `wxGTK32` to be compatible.

https://github.com/NixOS/nixpkgs/blob/f2bd8adf7b78d7616b52d0ef08865c7c2fcf189d/pkgs/applications/misc/prusa-slicer/default.nix#L147C46-L152

deftdawg commented:
> ```
> Cannot get default EGL display: EGL_BAD_PARAMETER
> Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
> Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
> Cannot get default EGL display: EGL_BAD_PARAMETER
> Cannot get default EGL display: EGL_BAD_PARAMETER
> Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
> Cannot create EGL sharing context: invalid display (last error: EGL_SUCCESS)
> Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
> Cannot create EGL context: invalid display (last error: EGL_SUCCESS)
> ```

I've upgrade to 23.11/Unstable (needed to move up to keep [Jovian-NixOS](https://github.com/Jovian-Experiments/Jovian-NixOS) working :smile: ) and I get the above as well with the window opening and then crashing... Put gdb on it and captured this backtrace of the segfault happening at `ImFont::CalcTextSizeA`.

```sh
echo -e "run\nbt\n\n" | LC_ALL=C gdb ./bambu-studio
```

```cpp
...
Thread 1 "bambustu_main" received signal SIGSEGV, Segmentation fault.
0x0000000001c8cf13 in ImFont::CalcTextSizeA(float, float, float, char const*, char const*, char const**) const ()
(gdb) bt
#0  0x0000000001c8cf13 in ImFont::CalcTextSizeA(float, float, float, char const*, char const*, char const**) const ()
#1  0x0000000001c42bd3 in ImGui::CalcTextSize(char const*, char const*, bool, float) ()
#2  0x0000000001699304 in Slic3r::GUI::NotificationManager::HintNotification::count_spaces() ()
#3  0x000000000169940b in Slic3r::GUI::NotificationManager::HintNotification::init() ()
#4  0x00000000016a279a in Slic3r::GUI::NotificationManager::HintNotification::retrieve_data(bool) ()
#5  0x0000000001456b27 in Slic3r::GUI::NotificationManager::push_hint_notification(bool) ()
#6  0x000000000110cc6b in Slic3r::GUI::GUI_App::post_init() ()
#7  0x000000000110f0fd in Slic3r::GUI::GUI_App::on_init_inner()::{lambda(wxIdleEvent&)#6}::operator()(wxIdleEvent&) const [clone .constprop.0] ()
#8  0x00007ffff0bbfd12 in wxEvtHandler::ProcessEventIfMatchesId(wxEventTableEntryBase const&, wxEvtHandler*, wxEvent&) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#9  0x00007ffff0bc0047 in wxEvtHandler::SearchDynamicEventTable(wxEvent&) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#10 0x00007ffff0bc01b0 in wxEvtHandler::TryHereOnly(wxEvent&) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#11 0x00007ffff0bc025a in wxEvtHandler::ProcessEventLocally(wxEvent&) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#12 0x00007ffff0bc0301 in wxEvtHandler::ProcessEvent(wxEvent&) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#13 0x00007ffff0aa29d2 in wxAppConsoleBase::ProcessIdle() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#14 0x00007ffff060e454 in wxAppBase::ProcessIdle() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_gtk3u_core-3.1.so.7
#15 0x00007ffff054cc9a in wxApp::DoIdle() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_gtk3u_core-3.1.so.7
#16 0x00007ffff054cd73 in wxapp_idle_callback () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_gtk3u_core-3.1.so.7
#17 0x00007fffede1cc4f in g_main_context_dispatch () from /nix/store/fr53rm643lg0rp4qmgzzpwsz5xxdqwk0-glib-2.76.3/lib/libglib-2.0.so.0
#18 0x00007fffede1d028 in g_main_context_iterate.constprop () from /nix/store/fr53rm643lg0rp4qmgzzpwsz5xxdqwk0-glib-2.76.3/lib/libglib-2.0.so.0
#19 0x00007fffede1d2df in g_main_loop_run () from /nix/store/fr53rm643lg0rp4qmgzzpwsz5xxdqwk0-glib-2.76.3/lib/libglib-2.0.so.0
#20 0x00007fffee605d75 in gtk_main () from /nix/store/9f63g8hd0f2iyxzpw41jrr9qqqhcfdab-gtk+3-3.24.38/lib/libgtk-3.so.0
#21 0x00007ffff055d8e5 in wxGUIEventLoop::DoRun() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_gtk3u_core-3.1.so.7
#22 0x00007ffff0acc6fd in wxEventLoopBase::Run() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#23 0x00007ffff0aa221b in wxAppConsoleBase::MainLoop() () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#24 0x00007ffff0b0d8f7 in wxEntry(int&, wchar_t**) () from /nix/store/rqqa7v39hzy4ppvjw9iqqsqr3i9g87q1-wxwidgets-3.1.7/lib/libwx_baseu-3.1.so.7
#25 0x00000000010d40ab in Slic3r::GUI::GUI_Run(Slic3r::GUI::GUI_InitParams&) ()
#26 0x000000000083b54b in Slic3r::CLI::run(int, char**) ()
#27 0x00000000007e04f4 in main ()
```

norpol commented on pkgs/top-level/all-packages.nix:
```diff
@@ -34698,6 +34698,18 @@ with pkgs;
+    wxGTK31 = wxGTK31.override {
+      withCurl = true;
+      withPrivateFonts = true;
+      withWebKit = true;
```
With `wxGTK32` `withWebKit=true;` will be default from now on

https://github.com/NixOS/nixpkgs/pull/268348/files

norpol commented:
@deftdawg Did you get it running already?

Seems to be a bug that was reported in PursaSlicer too.

https://github.com/prusa3d/PrusaSlicer/issues/9774

norpol commented:
After way too many weekends working on this I finally got it to compile:

* wxGTK31
* boost179
* bambu-studio `01.08.00.62`
* opvdb & tbb at 2021 


Here is my branch https://github.com/zhaofengli/nixpkgs/pull/2

deftdawg commented:
> @deftdawg Did you get it running already?

Hi @norpol, my home Internet's been offline all weekend, just got back online and tried with your PR merged into @zhaofengli fork.  Works 100%, thank you for pushing up that fix.  :beers:

Wulfsta commented:
Now that this is building, all checks are passing, and users are reporting that it's working, can a maintainer with permissions merge it? It would be great to have some movement on this since #239390, the package request for OrcaSlicer, depends on it.

deftdawg commented:
I think it's okay to merge in its current state, though it'd be good to have the edit filaments crash (mentioned above) fixed upstream.

That being said I can't see why merging orcaslicer would depend on this...

cab404 commented:
Will build and test in slicing, and then merge.

norpol commented:
OrcaSlicer builds and works with this current PR (just need to point the package to OrcaSlicer). OrcaSlicer already fixed the "Edit Filament" bug and as of latest beta release of BambuStudio they've also pulled in that fix (fix number 17 at the release notes).

I'd suggest to wait for the final 01.08 release and then finally merge.

https://github.com/bambulab/BambuStudio/issues/2605#issuecomment-1837613536

https://github.com/bambulab/BambuStudio/releases/tag/v01.08.01.57

cab404 commented:
> OrcaSlicer builds and works with this current PR (just need to point the package to OrcaSlicer). OrcaSlicer already fixed the "Edit Filament" bug and as of latest beta release of BambuStudio they've also pulled in that fix (fix number 17 at the release notes).
> 
> I'd suggest to wait for the final 01.08 release and then finally merge.
> 
> [bambulab/BambuStudio#2605 (comment)](https://github.com/bambulab/BambuStudio/issues/2605#issuecomment-1837613536)
> 
> https://github.com/bambulab/BambuStudio/releases/tag/v01.08.01.57

We will always be able to merge it afterwards.

cab404 commented:
Yup, everything builds and appears to be working in current state. Merging <3

cab404 commented:
> OrcaSlicer builds and works with this current PR (just need to point the package to OrcaSlicer). OrcaSlicer already fixed the "Edit Filament" bug and as of latest beta release of BambuStudio they've also pulled in that fix (fix number 17 at the release notes).
> 
> I'd suggest to wait for the final 01.08 release and then finally merge.
> 
> [bambulab/BambuStudio#2605 (comment)](https://github.com/bambulab/BambuStudio/issues/2605#issuecomment-1837613536)
> 
> https://github.com/bambulab/BambuStudio/releases/tag/v01.08.01.57

But yeah, edit filament is unfortunate

cab404 commented:
Will bump to 57, and then let's re-check it

cab404 commented:
https://github.com/NixOS/nixpkgs/pull/273350
